lib/FrontendTool/FrontendTool.cpp

```diff
@@ -237,12 +237,156 @@ template <> struct ObjectTraits<LoadedModuleTraceFormat> {
+    auto isSwiftinterface =
+      moduleFileType == file_types::TY_SwiftModuleInterfaceFile;
+
+    if (!(isSwiftmodule || isSwiftinterface))
+      continue;
+
+    auto dep = pathToModuleDecl.find(depPath);
+    if (dep != pathToModuleDecl.end()) {
+      // Great, we recognize the path! Check if the file is still around.
+
+      ModuleDecl *depMod = dep->second;
+      if(depMod->isResilient() && !isSwiftinterface) {
+        SmallString<256> moduleAdjacentInterfacePath(depPath);
+        computeAdjacentInterfacePath(moduleAdjacentInterfacePath);
+        if (fs::real_path(moduleAdjacentInterfacePath, buffer, true)) {
+          err << "The module " << depMod->getName().str() << " has library"
+              << " evolution enabled but we're recording a non-adjacent"
+              << " swiftmodule at\n" << depPath << "\nin the trace.";
+          llvm::report_fatal_error(err.str());
+        }
+        buffer.clear();
+      }
+
+      // FIXME: Better error handling
+      StringRef realDepPath
+        = fs::real_path(depPath, buffer, /*expand_tile*/true)
+        ? StringRef(depPath) // Couldn't find the canonical path, assume
+                             // this is good enough.
+        : buffer.str();
+      buffer.clear();
+
+      traceInfo.push_back({
+        /*Name=*/
+        depMod->getName(),
+        /*Path=*/
+        realDepPath,
+        // TODO: There is an edge case which is not handled here.
+        // When we build a framework using -import-underlying-module, or an
+        // app/test using -import-objc-header, we should look at the direct
+        // imports of the bridging modules, and mark those as our direct
+        // imports.
+        /*IsImportedDirectly=*/
+        importedModules.find(depMod) != importedModules.end(),
+        /*SupportsLibraryEvolution=*/
+        depMod->isResilient()
+      });
+
+      continue;
+    }
+
+    // If the depTracker had an interface, that means that we must've
+    // built a swiftmodule from that interface, so we should have that
+    // filename available.
+    if (isSwiftinterface) {
+      err << "Unexpected path for swiftinterface file:\n" << depPath << "\n";
+      errorUnexpectedPath();
+    }
+
+    // Skip cached modules in the prebuilt cache. We will add the corresponding
+    // swiftinterface from the SDK directly, but this isn't checked. :-/
+    if (StringRef(depPath).contains(prebuiltCachePath))
+      continue;
+
+    // If we have a swiftmodule next to an interface, that interface path will
+    // be saved (not checked), so don't save the path to this swiftmodule.
+    SmallString<256> moduleAdjacentInterfacePath(depPath);
+    computeAdjacentInterfacePath(moduleAdjacentInterfacePath);
+    if (pathToModuleDecl.find(moduleAdjacentInterfacePath)
+        != pathToModuleDecl.end())
+      continue;
+
+    if (!fs::real_path(moduleAdjacentInterfacePath, buffer,
+                       /*expand_tilde*/true)) {
```

Comment:
This isn't correct because it's not going through the VFS. I'm not sure what it's checking anyway; `fs::real_path` isn't the usual way to see if something exists (that'd be `fs::exists`).

Reply:
I see that it was already being used in the happy path and we should probably fix that in a follow-up.

Reply:
`exists` doesn't document what expansions it does, if any.

Reply:
Right, but you're not looking for expansions here. You're looking for an actual file next to where the swiftmodule was found.

Reply:
I don't understand what you mean. With `exists`, if I give it a path `~/path/to/Foo.swiftinterface`, then I don't know if it will expand the `~` or not before checking if the file exists or not.

Put another way, it is undocumented what `exists` does for non-absolute paths. `moduleAdjacentInterfacePath` may not be an absolute path, because `depPath` may not be an absolute path.

Reply:
You know that `depPath` is a path that got looked up in some way, and you're forming a path relative to that. You'd need to be consistent with however the dependency was found, and you don't know what logic that was.  Expanding tildes would probably be wrong, even.

Reply:
(Punted to later.)